### DISCUSSION TITLE
Button's TextBlock FontSize should inherit from the Button FontSize

### DISCUSSION THREAD

**Discussion by dlukach** [AvaloniaUI/Avalonia#7118] [Ideas]
**Is your feature request related to a problem? Please describe.**
Currently when defining a global deafult TextBlock style for FontSize, Button's which implement a fontsize are not able to overwrite the internal textblock's fontsize due to the style. Removing the FontSize global style fixes this currently. I believe this will be useful for projects which implement a default style theme like in my case where we have a default textblock size. 

**Describe the solution you'd like**
Make the fontsize of the Button's Textblock always depend on the Button font size instead of the TextBlock font size. 

**Describe alternatives you've considered**
Overwriting the content presenter style of our buttons to make this work or disabling a global font size theme. 

**Additional context**
This was discussed in Issue #4353 and I left a comment there

**Comment by maxkatz6** [comment]
If you set global textblock style, you can also set global button style template-binding to the local button.fontsize value

**Comment by dlukach** [comment]
Surprisingly I set my global button style to Textblock.FontSize = {Templatebinding FontSize} & it didn't work.

**Reply by maxkatz6** [reply]
```xaml
<Style Selector="TextBlock">
   <Setter Property="FontSize" Value="12" />
</Style>
```
```xaml
<Style Selector="Button TextBlock">
   <Setter Property="FontSize" Value="{Binding $parent[Button].FontSize}" />
</Style>
```
Or maybe this will work, need to test
```xaml
<Style Selector="Button TextBlock">
   <Setter Property="FontSize" Value="{TemplateBinding FontSize}" />
</Style>
```

**Reply by dlukach** [reply]
I've tried these and it didn't work. The problem is the TextBlock has a global font style that will ignore these.

**Reply by maxkatz6** [reply]
> The problem is the TextBlock has a global font style that will ignore these

Not exactly right. See my another answer about prioritites.

**Comment by maxkatz6** [comment]
Basically reason of this behavior is simple after knowing how style system works + knowing that Avalonia has inherited properties. 

How font size is resolved going by priority from lowest:
1. Font size property default value
2. Inherited value. When you set TextBlock.FontSize property on parent control, each TextBlock children will inherit same value. Technically this value was not set on text block, it doesn't have high priority. 
3. TemplateBinding values set directly on TextBlock.
4. Style values defined with style setters set directly on TextBlock using styles.
5. StyleTrigger values defined with style setters with trigger pseudoclasses (pointerover, pressed...) set directly on TextBlock using styles.
6. Local values set directly on the control without styles (i.e. set on specific control instance). 
7. Animation values always have the highest priority.

**Reply by maxkatz6** [reply]
And in terms of custom themes. They shouldn't set text block font styles **globally**, because it will unavoidable break inherited values. 

Instead if they need to set font size globally, they should change it in the inheritance root (Window), like we do in our themes. See https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Themes.Fluent/Controls/Window.xaml#L5

**Reply by robloo** [reply]
Is this priority documented? It's a great list

**Reply by maxkatz6** [reply]
I don't remember if there is a specific docs page, but it's basically defined by this enum: http://reference.avaloniaui.net/api/Avalonia.Data/BindingPriority/

Where "Unset" means not only "no value, use default value" but also "no value, try to get inherited from parent or use default otherwise".

**Reply by maxkatz6** [reply]
These pages are helpful in this context:
https://docs.avaloniaui.net/docs/authoring-controls/defining-properties#direct-avaloniaproperties
https://docs.avaloniaui.net/docs/styling/troubleshooting
